### ISSUE TITLE
Consider serializing `Psbt` using base 64

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3454]
In #2906 we had to patch the `serde` stuff for `Psbt` but its still ugly. We could just serde using base64 but note that might make features a bit clunky, one would then have to enable both `serde` and `base64` to get `serde` stuff. 

Probably needs a bit of playing with and a bit of discussion. Not super high priority but we should probably pin this down at some stage - note breaking serde serialization can be painful for upgraders but for `Psbt` this is probably ok, or do folk keep PSBTs laying around between releases?

**Comment by apoelstra** [comment]
If nobody is keeping PSBTs laying around between releases we could just drop the serde impl entirely. For normal ephemeral usage they should be just using the standard base64 encoding.

**Comment by tcharding** [comment]
So if/when we close this we should add migration code also, right?

**Comment by apoelstra** [comment]
Whew. Yeah, we definitely should. But that may be a real pain in the ass for something as big as `Psbt` :(

**Comment by tcharding** [comment]
Non-critical but I don't know how you are pronouncing the word 'whew' or exactly what that means (I looked up whew vs phew). Is this how you are using it: https://duckduckgo.com/?t=ffab&q=pronounce+whew&ia=web

**Comment by tcharding** [comment]
Kind of funny, we are culturally (Australia/Canada) so similar but some little things are totally foreign :)

**Comment by apoelstra** [comment]
I am pronouncing it like the "US" tab in https://dictionary.cambridge.org/us/pronunciation/english/whew

I think "phew" is a pretty good approximation to what I mean by it, though I think "phew" connotes relief, which is not what I mean usually. Usually the closest emoji is :sweat_smile:.

**Comment by tcharding** [comment]
Interesting, from listening to that I would have spelled it "phew" and understood you the same if I heard you say it. 

>  I think "phew" connotes relief, which is not what I mean usually.

You mean it as if you accompanied the expression with a mock wipe of the brow, right?

**Comment by apoelstra** [comment]
>You mean it as if you accompanied the expression with a mock wipe of the brow, right?

Yep. There is a `sweat_smile` emoji which is pretty close but Github won't render it..

**Comment by sanket1729** [comment]
> If nobody is keeping PSBTs laying around between releases we could just drop the serde impl entirely. For normal ephemeral usage they should be just using the standard base64 encoding.

The serde implementation is particularly valuable for CLI tools like hal, which use it for reading and writing JSON data. Don't care much about breaking it, but I would like to have serde support if possible

**Comment by Ademan** [comment]
At a glance, to me, it looks like hal doesn't actually rely on rust-bitcoin's Psbt serde implementation anyway. https://github.com/stevenroose/hal/blob/master/src/psbt.rs

I'm not certain what other users or use cases there might be out there, but migration is definitely important.

@tcharding It does seem like the Psbt serde implementation gets exercised here https://github.com/rust-bitcoin/corepc/blob/master/client/src/client_sync/v17/raw_transactions.rs#L95 and presumably its deserialization needs to match core's output, but I understand `corepc-client` isn't really for general consumption.

I'm not sure what the right thing to do for corepc-client is, but we could put a compatibility newtype in a feature that maintains the current de/serialization, I'm just not sure if/how to keep people from gunning down their feet other than adding it to the changelog and hoping people see it.

We could actually put all serde de/serialization in newtypes, so code that expects `Psbt` to implement `Serialize` and `Deserialize` will fail to build, but `SerdePsbt(Psbt)` (or whatever) will work, as will `LegacyPsbtSerde(Psbt)`

**Comment by Ademan** [comment]
That was a lot of thinking out loud, but here's a summary and a proposal:

Hide `Psbt` serde behind two features

* `serde-psbt` enables building `bitcoin::serde_utils::psbt`
* `serde-psbt-legacy` enables building `bitcoin::serde_utils::psbt_legacy` which matches current behavior

Then, `Serialize` and `Deserialize` traits for `Psbt` are implemented depending on the two features.

|                             | serde-psbt-legacy: No | serde-psbt-legacy: Yes |
| ---------------------- | --------------------------------| ----------------------------------|
| serde-psbt: No | No `Psbt` serde traits  | `Psbt` serde is `bitcoin::serde_utils::psbt_legacy` |
| ---------------------- | --------------------------------| ----------------------------------|
| serde-psbt: Yes | `Psbt` serde is `psbt::serde_utils::psbt` | No `Psbt` serde traits, must manually use `serde(with = ...)` |

This way, a user naively upgrading, having neither `serde-psbt` nor `serde-psbt-legacy` enabled will get a compilation error, and look at the changelog (hopefully), a user who knows exactly what they want can use it trivially, and a user that needs to migrate can use both simultaneously.

**Comment by tcharding** [comment]
>  It does seem like the Psbt serde implementation gets exercised here https://github.com/rust-bitcoin/corepc/blob/master/client/src/client_sync/v17/raw_transactions.rs#L95 and presumably its deserialization needs to match core's output, but

That is not using `serde` because there is no `Psbt` type, its just a string.

>  I understand corepc-client isn't really for general consumption.

It is for general consumption, limited though to writing integration tests.

> Hide Psbt serde behind two features

Thanks for thinking about this problem. However I personally think we should just break the current `serde` impls and downstream can use `0.32.x` to deserialize their types then reserialize them with `0.33`. Then they can just yell at me in person and I'll shrug and say "release a conversion crate to help others and quit whinging" :)
